src/analyses/deadlock.ml

```diff
@@ -4,109 +4,94 @@ open Prelude.Ana
-        let lockAddrs = eval_exp_addr (Analyses.ask_of_ctx ctx) (List.hd arglist) in
-        if List.compare_length_with lockAddrs 1 = 0 then
-          let inLockAddrs e = List.exists (fun r -> ValueDomain.Addr.equal r e.addr) lockAddrs in
-          D.filter (neg inLockAddrs) ctx.local
-        else ctx.local
-      | _ -> ctx.local
-
+  let event ctx (e: Events.t) octx =
+    match e with
+    | Lock addr ->
+      let after = (addr, ctx.prev_node) in
+      D.iter (fun before ->
+          side_lock_event_pair ctx before after
+        ) ctx.local;
+      D.add after ctx.local
+    | Unlock addr ->
+      let inLockAddrs (e, _) = Lock.equal addr e in
+      D.filter (neg inLockAddrs) ctx.local
+    | _ ->
+      ctx.local
+
+  let query ctx (type a) (q: a Queries.t): a Queries.result =
+    match q with
+    | WarnGlobal g ->
+      let g: V.t = Obj.obj g in
+
+      let module LH = Hashtbl.Make (Lock) in
+      let module LS = Set.Make (Lock) in
+      (* TODO: find all cycles/SCCs *)
+      let global_visited_locks = LH.create 100 in
+
+      (* DFS *)
+      let rec iter_lock (path_visited_locks: LS.t) (path_visited_lock_event_pairs: LockEventPair.t list) (lock: Lock.t) =
+        if LS.mem lock path_visited_locks then (
+          (* normalize path_visited_lock_event_pairs such that we don't get the same cycle multiple times, starting from different events *)
+          let min = List.min ~cmp:LockEventPair.compare path_visited_lock_event_pairs in
+          let (mini, _) = List.findi (fun i x -> LockEventPair.equal min x) path_visited_lock_event_pairs in
+          let (init, tail) = List.split_at mini path_visited_lock_event_pairs in
+          let normalized = List.rev_append init (List.rev tail) in (* backwards to get correct printout order *)
+          let msgs = List.concat_map (fun ((before_lock, before_node), (after_lock, after_node)) ->
+              [
+                (Pretty.dprintf "lock before: %a" Lock.pretty before_lock, Some (UpdateCil.getLoc before_node));
+                (Pretty.dprintf "lock after: %a" Lock.pretty after_lock, Some (UpdateCil.getLoc after_node));
+              ]
+            ) normalized
+          in
+          M.msg_group Warning ~category:Deadlock "Locking order cycle" msgs
+        )
+        else if not (LH.mem global_visited_locks lock) then begin
+          LH.replace global_visited_locks lock ();
+          let new_path_visited_locks = LS.add lock path_visited_locks in
+          G.iter (fun to_lock lock_event_pairs ->
+              MayLockEventPairs.iter (fun lock_event_pair ->
+                  let new_path_visited_lock_event_pairs' = lock_event_pair :: path_visited_lock_event_pairs in
+                  iter_lock new_path_visited_locks new_path_visited_lock_event_pairs' to_lock
+                ) lock_event_pairs
+            ) (ctx.global lock)
+        end
+      in
+
+      Stats.time "deadlock" (iter_lock LS.empty []) g
+    | _ -> Queries.Result.top q
 end
 
 let _ =
```

Comment:
Here, we should probably record a dependency on `mutex`.
If it is not enabled, the `Lock` and `Unlock` events won't be emitted, and we wrongly report no deadlocks.

Reply:
True, this analysis is useless otherwise, so I'll add the dependency.